apps/predictions/lib/predictions_pub_sub.ex

```diff
@@ -32,12 +36,15 @@ defmodule Predictions.PredictionsPubSub do
     )
   end
 
-  @spec subscribe(Route.id_t()) :: [Prediction.t()]
-  @spec subscribe(Route.id_t(), GenServer.server()) :: [Prediction.t()]
-  def subscribe(route_id, server \\ __MODULE__) do
-    StreamSupervisor.ensure_stream_is_started(route_id)
-    {registry_key, predictions} = GenServer.call(server, {:subscribe, route_id})
-    Registry.register(:prediction_subscriptions_registry, registry_key, route_id)
+  @spec subscribe(Route.id_t(), Stop.id_t(), direction_id_t()) :: [Prediction.t()]
+  @spec subscribe(Route.id_t(), Stop.id_t(), direction_id_t(), GenServer.server()) :: [
+          Prediction.t()
+        ]
+  def subscribe(route_id, stop_id, direction_id, server \\ __MODULE__) do
+    key = prediction_key(route_id, stop_id, direction_id)
+    StreamSupervisor.ensure_stream_is_started(key)
+    {registry_key, predictions} = GenServer.call(server, {:subscribe, key})
+    Registry.register(:prediction_subscriptions_registry, registry_key, key)
```

Comment:
suggestion (non-blocking): If you wanted to, you could store predictions by route+stop+direction, while leaving the streaming as by route. In other words, the server-client contract could be based on route+stop+direction, while the server-API client would just be based on route. The con here would be that you'd be streaming more data than you need from the API, but the pro would be less churn in starting small streams. I'm not sure what the performance implications would be in practice, but my gut instinct is that there are very likely to be multiple users looking at the same route, even if they are looking at different stops along it.

Reply:
@arkadyan I ended up doing this!